File: python/src/code_smells/dispensables/lazy_class.py
"""
Lazy Class Code Smell Example

A class that doesn't do enough to justify its existence.
It has minimal functionality and could be inlined or removed.

Refactoring suggestion: Inline Class, Collapse Hierarchy
"""
from typing import TypedDict


class Address(TypedDict):
    name: str
    line1: str
    city: str | None


class ShippingLabelBuilder:
    """
    Lazy class: only has one trivial method.

    This class adds no value - its functionality could be
    a simple function or moved elsewhere.
    """

    def build(self, address: Address) -> str:
        city_part = f", {address['city']}" if address.get('city') else ''
        return f"{address['name']} — {address['line1']}{city_part}"


def print_shipping_label() -> None:
    address: Address = {
        'name': 'John Doe',
        'line1': '123 Main St',
        'city': 'New York',
    }

    label_builder = ShippingLabelBuilder()
    label = label_builder.build(address)
    print(label)

File: go/code_smells/dispensables/lazy_class.go
package dispensables

import "fmt"

// Code smell: Lazy Class [Clase perezosa].
// La clase ShippingLabelBuilder solo tiene un método que realiza una concatenación
// de strings simple, algo que podría resolverse con una función pura.

// Ejercicio: Reescribe el código para eliminar la necesidad de la clase ShippingLabelBuilder.

// Mantener una estructura de clase para una lógica tan simple te obliga a instanciar objetos
// innecesariamente y añade capas de abstracción que dificultan la legibilidad del código
// sin ofrecer beneficios a cambio.

type Address struct {
	Name  string
	Line1 string
	City  *string
}

type ShippingLabelBuilder struct{}

func (s *ShippingLabelBuilder) Build(a Address) string {
	cityStr := ""
	if a.City != nil {
		cityStr = ", " + *a.City
	}
	return fmt.Sprintf("%s — %s%s", a.Name, a.Line1, cityStr)
}

func PrintShippingLabel() {
	city := "New York"
	address := Address{
		Name:  "John Doe",
		Line1: "123 Main St",
		City:  &city,
	}

	labelBuilder := &ShippingLabelBuilder{}
	label := labelBuilder.Build(address)
	fmt.Println(label)
}
